README.md
smart-mode-line [![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=GLZLRLW72Q8G2)
---------------

Smart Mode Line is a sexy mode-line for Emacs. It aims to be easy to
read from small to large monitors by using *colors*, a *prefix feature*, and
*smart truncation*. 

[New in v2.5](https://github.com/Bruce-Connor/smart-mode-line/blob/master/news.md)
===========
- Emacs 24.4 compatible.
- Integration with [Projectile](https://github.com/bbatsov/projectile)!
- Display `current-directory` in Shell and eshell.
- New value for `sml/theme`: `automatic` (highly recommended).
- `sml/apply-theme` is interactive and has completion.
- Smart-mode-line themes are now regular themes.

[Further News](https://github.com/Bruce-Connor/smart-mode-line/blob/master/news.md)

Images
======
**Dark Theme**  
![Dark Theme Screenshot](https://raw.github.com/Bruce-Connor/smart-mode-line/master/screenshot-2013-11-11-dark.png)

**Light Theme**  
![Light Theme Screenshot](https://raw.github.com/Bruce-Connor/smart-mode-line/master/screenshot-2013-11-11-light.png)

Installation
===
**smart-mode-line** is available on Melpa, and that's the recommended
way of installing it. If you do that, you can simply activate it with:

    (sml/setup)

To install it manually, you need **emacs-version >= 24.3**. First
make sure you install [dash.el](https://github.com/magnars/dash.el)
(which is a dependency), then make sure *"smart-mode-line.el"* is in
your load path, and finally place this code in your `.emacs` file:

    (require 'smart-mode-line)
    (sml/setup)

To change the color theme, do one of the following:

    (sml/apply-theme 'dark)
    (sml/apply-theme 'light)
    (sml/apply-theme 'respectful)
    (sml/apply-theme 'automatic)

#### Instalation Issues (FAQ) ####

- **Problem:** If emacs keeps warning you that *"Loading themes can run
lisp code"* and asking *"Would you like to mark this theme as safe for
future sessions?"*. That is probably an issue with your `init.el` or
`.emacs` file.
- **Solution:** Make sure the `(custom-set-variables ...)` sexp is at
  the very top of your `.emacs` file. That is the right place for it.
  If that doesn't work, you can work around it with the code `(setq sml/no-confirm-load-theme t)`,
  but we recommend you try to figure out what's wrong with your configs.
  
Features
===
Its main features include:

 1. **Color coded**:  
    Highlights the most important information for you
    (buffer name, modified state, line number). Don't
    like the colors? See item *5.*!

 2. **Fixed width** (if you want):  
    Lets you set a maxium width for the path name and mode names, and
    truncates them intelligently (truncates the directory, not the
    buffer name). Also let's you **right indent** strings in the
    mode-line (see `sml/mode-width`).

 3. **Directory as Prefixes**:  
    Prefix feature saves a LOT of space. e.g. **"~/.emacs.d/"**
    is translated to **":ED:"** in the path (open a file inside
    this folder to see it in action). Long path names you
    are commmonly working on are displayed as short
    abbreviations. Set your own prefixes to make best use
    of it (by configuring `sml/replacer-regexp-list`). Mousing
    over the abbreviated path will show you the full
    path. See below for examples.  	

 4. **Hide or Highlight minor-modes**:  
    The [rich-minority](https://github.com/Bruce-Connor/rich-minority)
    package saves even more space. Select which minor modes you don't
    want to see listed by adding them to the variable
    `rm-excluded-modes`, or even highlight the modes that are more
    important with the variable `rm-text-properties`. This will filter
    out the modes you don't care about and unclutter the modes list
    (mousing over the modes list still shows the full list).

 5. **Very easy to configure**:  
    All colors and variables are customizable. You can change the
    whole theme with `sml/apply-theme`, or just customize anything
    manually with `sml/customize` and `sml/customize-faces`. There are
    *DOZENS* of variables to customize your mode-line, just pop over
    there and have a look!

 6. **Compatible with absolutely anything**:  
    I'm serious. Versions 2.0 and above should be compatible with
    **any** other packages that display information in the mode-line
    (evil, nyan-mode, elscreen, display-battery-mode, etc). If you
    find *ANYTHING* that does not appear as it should, file a bug report
    and I'll get to it.
    
Important Variables:
===
All variables can be edited by running `sml/customize`, and the
documentations are mostly self explanatory, I list here only the
most important ones.

 1. `sml/theme`  
  Choose what theme you want to use for the mode-line colors. For now
  there are 3 different themes: `dark`, `light`, and `respectful`.
 
 1. `sml/shorten-directory` and `sml/shorten-modes`  
  Setting both of these to `t` garantees a fixed width mode-line
  (directory name and minor-modes list will be truncated to fit). To
  actually define the width, see below.
  
 2. `sml/name-width` and `sml/mode-width`  
  Customize these according to the width of your emacs frame. I set
  them to `40` and `full` respectively, and the mode-line fits
  perfectly when the frame is split in two even on my laptop's small
  17" monitor. `full` means everything after the minor-modes will be
  right-indented.
  
 3. `sml/replacer-regexp-list`  
  This variable is a list of (REGEXP REPLACEMENT) that is used
  to parse the path. The replacements are applied
  sequentially. This allows you to greatly abbreviate the path
  that's shown in the mode-line. If this abbreviation is of
  the form **":SOMETHING:"**, it is considered a prefix and get's
  a different color (you can change what's considered a prefix
  by customizing `sml/prefix-regexp`).  
  For example, if you do a lot of work on a folder called
  **"~/Dropbox/Projects/In-Development/"** almost half the
  mode-line would be occupied just by the folder name, which
  is much less important than the buffer name. But, you can't
  just hide the folder name, since editting a file in
  **"~/Dropbox/Projects/In-Development/Source"** is VERY different
  from editting a file in **"~/Dropbox/Projects/Source"**. By
  setting up a prefix for your commonly used folders, you get
  all that information without wasting all that space. In this
  example you could set the replacement to **":ProjDev:"** or just
  **":InDev:"**, so the path shown in the mode-line will be
  **":ProjDev:Source/"** (saves a lot of space without hiding
  information).  

Here go some more useful examples:

    (add-to-list 'sml/replacer-regexp-list '("^~/Dropbox/Projects/In-Development/" ":ProjDev:") t)
    (add-to-list 'sml/replacer-regexp-list '("^~/Documents/Work/" ":Work:") t)
    
    ;; Added in the right order, they even work sequentially:
    (add-to-list 'sml/replacer-regexp-list '("^~/Dropbox/" ":DB:") t)
    (add-to-list 'sml/replacer-regexp-list '("^:DB:Documents" ":DDocs:") t)
    (add-to-list 'sml/replacer-regexp-list '("^~/Git-Projects/" ":Git:") t)
    (add-to-list 'sml/replacer-regexp-list '("^:Git:\\(.*\\)/src/main/java/" ":G/\\1/SMJ:") t)


news.md
New in v2.6
===========
- Filtering of minor modes is now delegated to the
  [rich-minority](https://github.com/Bruce-Connor/rich-minority)
  package. This has 3 effects:
  - The original (black list) method of filtering is given by the
    variable `rm-blacklist`. `sml/hidden-modes` still works, but
    is obsolete.
  - Alternatively, modes can be filtered by a whitelist, using `rm-whitelist`.
  - Modes can also be propertized, using `rm-text-properties`.

- The file-path display can have different minimum and maximum widths.
  Just set `sml/name-width` to a cons cell `(MIN . MAX)`.

New in v2.5
===========
- Emacs 24.4 compatible.
- Integration with [Projectile](https://github.com/bbatsov/projectile)!
- Display `current-directory` in Shell and eshell.
- New value for `sml/theme`: `automatic` (highly recommended).
- `sml/apply-theme` is interactive and has completion.
- Smart-mode-line themes are now regular themes, which makes it easier to define new ones. You can activate them with `(load-theme 'smart-mode-line-light)`, though `(sml/apply-theme 'light)` still works and has the advantages of disabling the previous theme.

New in v2.1
===========
- **Themes!** (yes, finally). Besides the usual dark theme, we now have
light and a respectful theme. See `sml/theme`.
- **Zero Compromise.** This means you don't have to
give anything up in order to use this package. Every single bit of
information which is displayed in the standard mode-line is present in
`sml` (though some might need to be turned on). This includes 100%
compatibility with other external packages which display information
in the mode-line.
- Performance improvements.



smart-mode-line-dark-theme.el
;;; smart-mode-line-dark-theme.el --- Dark theme for smart-mode-line

;; Copyright (C) 2014 Artur Malabarba <bruce.connor.am@gmail.com>

;; Author: Artur Malabarba <bruce.connor.am@gmail.com>
;; URL: http://github.com/Bruce-Connor/smart-mode-line
;; Version: 0.1a
;; Separator: -

;;; License:
;;
;; This file is NOT part of GNU Emacs.
;;
;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 

;;; Change Log:
;; 0.1a - 2014/05/14 - Created File.
;;; Code:

(deftheme smart-mode-line-dark "Dark theme for smart-mode-line.")

(custom-theme-set-faces
 'smart-mode-line-dark
 '(mode-line-buffer-id ((t :inherit sml/filename :foreground nil :background nil))) 
 '(mode-line-inactive ((t :foreground "gray60" :background "#404045" :inverse-video nil)))
 '(mode-line     ((t :foreground "gray60" :background "black" :inverse-video nil)))
 '(sml/global    ((t :foreground "gray50" :inverse-video nil)))
 '(sml/modes     ((t :inherit sml/global :foreground "White")))
 '(sml/filename  ((t :inherit sml/global :foreground "#eab700" :weight bold)))
 '(sml/prefix    ((t :inherit sml/global :foreground "#bf6000")))
 '(sml/read-only ((t :inherit sml/not-modified :foreground "DeepSkyBlue")))
 '(persp-selected-face ((t :foreground "ForestGreen" :inherit sml/filename)))
 '(helm-candidate-number ((t :foreground nil :background nil :inherit sml/filename))))

;;;###autoload
(when load-file-name
  (add-to-list 'custom-theme-load-path
               (file-name-as-directory (file-name-directory load-file-name))))

(provide-theme 'smart-mode-line-dark)
;;; smart-mode-line-dark-theme.el ends here.


smart-mode-line-light-theme.el
;;; smart-mode-line-light-theme.el --- Light theme for smart-mode-line

;; Copyright (C) 2014 Artur Malabarba <bruce.connor.am@gmail.com>

;; Author: Artur Malabarba <bruce.connor.am@gmail.com>
;; URL: http://github.com/Bruce-Connor/smart-mode-line
;; Version: 0.1a
;; Separator: -

;;; License:
;;
;; This file is NOT part of GNU Emacs.
;;
;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 

;;; Change Log:
;; 0.1a - 2014/05/14 - Created File.
;;; Code:

(deftheme smart-mode-line-light "Light theme for smart-mode-line.")

(custom-theme-set-faces
 'smart-mode-line-light
 '(mode-line-buffer-id ((t :inherit sml/filename :foreground nil :background nil))) 
 '(mode-line-inactive ((t :foreground "grey20" :background "#fdf6e3" :inverse-video nil)))
 '(mode-line     ((t :foreground "black" :background "grey85" :inverse-video nil)))
 '(sml/global    ((t :foreground "gray20" :inverse-video nil)))
 '(sml/modes     ((t :inherit sml/global :foreground "Black")))
 '(sml/filename  ((t :inherit sml/global :foreground "Blue"      :weight bold)))
 '(sml/prefix    ((t :inherit sml/global :foreground "#5b2507"   :weight bold)))
 '(sml/read-only ((t :inherit sml/not-modified :foreground "DarkGreen" :weight bold))))

;;;###autoload
(when load-file-name
  (add-to-list 'custom-theme-load-path
               (file-name-as-directory (file-name-directory load-file-name))))

(provide-theme 'smart-mode-line-light)
;;; smart-mode-line-light-theme.el ends here.


smart-mode-line-respectful-theme.el
;;; smart-mode-line-respectful-theme.el --- Respectful theme for smart-mode-line

;; Copyright (C) 2014 Artur Malabarba <bruce.connor.am@gmail.com>

;; Author: Artur Malabarba <bruce.connor.am@gmail.com>
;; URL: http://github.com/Bruce-Connor/smart-mode-line
;; Version: 0.1a
;; Separator: -

;;; License:
;;
;; This file is NOT part of GNU Emacs.
;;
;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 

;;; Change Log:
;; 0.1a - 2014/05/14 - Created File.
;;; Code:

(deftheme smart-mode-line-respectful
  "Respectful theme for smart-mode-line.
Tries to respect the colors chosen by your global theme.
Results may vary.")

(custom-theme-set-faces
 'smart-mode-line-respectful
 '(sml/global    ((t :inherit font-lock-preprocessor-face)))
 '(sml/filename  ((t :inherit mode-line-buffer-id)))
 '(sml/prefix    ((t :inherit (font-lock-variable-name-face sml/global))))
 '(sml/read-only ((t :inherit (font-lock-type-face sml/not-modified))))
 '(sml/modes     ((t :foreground nil :inherit sml/filename :weight normal))))

;;;###autoload
(when load-file-name
  (add-to-list 'custom-theme-load-path
               (file-name-as-directory (file-name-directory load-file-name))))

(provide-theme 'smart-mode-line-respectful)
;;; smart-mode-line-respectful-theme.el ends here.


themes/smart-mode-line-powerline-theme.el
;;; smart-mode-line-powerline-theme.el --- smart-mode-line theme that mimics the powerline appearance.

;; Copyright (C) 2014 Artur Malabarba <bruce.connor.am@gmail.com>

;; Author: Artur Malabarba <bruce.connor.am@gmail.com>
;; URL: http://github.com/Bruce-Connor/smart-mode-line
;; Version: 0.1a
;; Package-Requires: ((emacs "24.3") (powerline "2.3") (smart-mode-line "2.5"))
;; Separator: -

;;; License:
;;
;; This file is NOT part of GNU Emacs.
;;
;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;; 

;;; Change Log:
;; 0.1a - 2014/05/15 - powerline-theme essentially finished.
;; 0.1a - 2014/05/14 - Created File.
;;; Code:

(deftheme smart-mode-line-powerline
  "Powerline theme for smart-mode-line.
Mimics the appearance of powerline.")

(require 'powerline)

(let ((l0 "black")
      (l3 (or (face-background 'powerline-active1) "Grey30"))
      (l8 (or (face-background 'powerline-active2) "Grey80"))
      (separator-left 
       '(intern (format "powerline-%s-%s"
                        powerline-default-separator
                        (car powerline-default-separator-dir))))
      (separator-right 
       '(intern (format "powerline-%s-%s"
                        powerline-default-separator
                        (cdr powerline-default-separator-dir)))))
  (custom-theme-set-faces
   'smart-mode-line-powerline
   `(mode-line-buffer-id ((t :inherit sml/filename :foreground nil :background nil))) 
   `(mode-line-inactive ((((background dark)) :foreground "gray60" :background "Black"
                          :slant italic :box (:line-width -3 :color "black"))
                         (((background light)) :foreground "gray60" :background "Black"
                          :slant italic :box (:line-width -2 :color "white"))))
   `(mode-line     ((t :foreground "gray60" :background "black" :box (:line-width -1 :color "Black"))))
   `(sml/global    ((t :foreground "gray50" :inverse-video nil)))
   
   ;; Layer 0
   `(sml/line-number         ((t :foreground "White" :inherit sml/global :weight bold :background ,l0)))
   `(sml/remote              ((t :inherit sml/global :background ,l0)))
   `(sml/col-number          ((t :inherit sml/global :background ,l0)))
   `(sml/numbers-separator   ((t :inherit sml/col-number :background ,l0)))
   `(sml/client              ((t :inherit sml/prefix :background ,l0)))
   `(sml/mule-info           ((t :inherit sml/global :background ,l0)))
   `(sml/not-modified        ((t :inherit sml/global :background ,l0)))
   '(sml/read-only           ((t :inherit sml/not-modified :foreground "Cyan")))

   ;; 3
   `(sml/prefix    ((t :background ,l3 :inherit sml/global :foreground "#bf6000")))
   `(sml/filename  ((t :background ,l3 :inherit sml/global :foreground "gold" :weight bold)))
   `(sml/sudo      ((t :background ,l3 :inherit sml/outside-modified)))
   `(sml/git       ((t :background ,l3 :inherit (sml/read-only sml/prefix))))
   `(sml/folder    ((t :background ,l3 :inherit sml/global :weight normal :foreground "Black")))
   
   ;; 8
   `(sml/name-filling        ((t :background ,l8 :inherit sml/prefix :weight normal)))
   `(sml/position-percentage ((t :background ,l8 :inherit sml/prefix :weight normal :foreground "#330000")))
   `(sml/modes               ((t :background ,l8 :inherit sml/global :foreground "Black")))
   `(sml/process             ((t :background ,l8 :inherit sml/prefix)))
   `(sml/vc                  ((t :background ,l8 :inherit sml/git :foreground "#0000aa")))
   `(sml/vc-edited           ((t :background ,l8 :inherit sml/prefix :foreground "#330000")))

   ;; 3
   ;; minor modes

   ;; 0
   `(sml/discharging         ((t :background ,l0 :inherit sml/global :foreground "Red")))
   `(sml/time                ((t :background ,l0 :inherit sml/modes)))

   `(persp-selected-face ((t :foreground "ForestGreen" :inherit sml/filename)))
   `(helm-candidate-number ((t :foreground nil :background nil :inherit sml/filename))))
  (custom-theme-set-variables
   'smart-mode-line-powerline
   '(sml/mode-width (if (eq powerline-default-separator 'arrow) 'right 'full))
   `(sml/pre-id-separator
     '(:eval (propertize " " 'display (funcall ,separator-left nil 'powerline-active1))))
   `(sml/pos-id-separator
     '(:eval (propertize " " 'display (funcall ,separator-left 'powerline-active1 'powerline-active2))))
   `(sml/pre-minor-modes-separator
     '(:eval (propertize " " 'display (funcall ,separator-right 'powerline-active2 'powerline-active1))))
   `(sml/pos-minor-modes-separator
     '(:eval (propertize " " 'display (funcall ,separator-right 'powerline-active1 nil))))
   '(sml/pre-modes-separator
     (propertize " " 'face 'sml/modes))))

;;;###autoload
(when load-file-name
  (add-to-list 'custom-theme-load-path
               (file-name-as-directory (file-name-directory load-file-name))))

(provide-theme 'smart-mode-line-powerline)
;;; smart-mode-line-powerline-theme.el ends here.
